### PR TITLE
Be less global

### PR DESCRIPTION
**Pull request by islemaster**
That is: Take advantage of [`browserify-global-shim`](https://github.com/rluba/browserify-global-shim) and stop treating `React`, `ReactDOM` and `Radium` as globals.

Apps [already global-shims these three modules](https://github.com/code-dot-org/code-dot-org/blob/staging/apps/package.json#L33) which means in our build, requires for those things get transformed to retrieve them from window instead of loading the module.  We do this [because `code-studio` already contains a copy of these and puts them on window](https://github.com/code-dot-org/code-dot-org/blob/staging/code-studio/src/js/code-studio.js#L18).

However, in our tests and in the applab-api build we don't use `browserify-global-shim`, so in those builds requires for React, etc. will actually load the module.  This is good because it lets us undo some kludge to set up globals in those contexts.

All this means we should be requiring React, ReactDOM and Radium in apps as if there was nothing special about them.  That's what this PR does.

### CODE REVIEW

**Comment by islemaster** [comment]
FYI @pcardune @Bjvanminnen

File: .eslintrc.js [hunk 1]
```diff
@@ -20,9 +20,7 @@ module.exports = {
   },
   "globals": {
     "$": true,
-    "jQuery": true,
-    "React": true,
-    "ReactDOM": true
+    "jQuery": true
```
Comment (islemaster) [inline .eslintrc.js:23]: I'd _really_ like to do this for jQuery too, but that's a big enough task that I'm saving it for another PR.

File: apps/src/applab/api-entry.js [hunk 2]
```diff
@@ -2,10 +2,7 @@
 
 // third party dependencies that are provided as globals in code-studio but
 // which need to be explicitly required here.
-window.React = require('react');
-window.ReactDOM = require('react-dom');
 window.$ = require('jquery');
-window.Radium = require('radium');
```
Comment (islemaster) [inline apps/src/applab/api-entry.js:8]: Here's one of the places we no longer need to put these on the global namespace :tada:

File: apps/src/applab/designMode.js [hunk 3]
```diff
@@ -1,7 +1,7 @@
 /* global Applab, dashboard */
 
-// TODO (brent) - make it so that we dont need to specify .jsx. This currently
-// works in our grunt build, but not in tests
```
Comment (islemaster) [inline apps/src/applab/designMode.js:4]: Pretty sure this TODO is outdated now.

File: apps/src/package.json [hunk 4]
```diff
@@ -11,7 +11,6 @@
     ]
   },
   "browserify-shim": {
-    "marked": "global:marked",
-    "radium": "global:Radium"
+    "marked": "global:marked"
```
Comment (islemaster) [inline apps/src/package.json:14]: This was a little tricky; the `browserify-shim` entry for Radium here is now redundant with the `browserify-global-shim` entry in apps/package.json, and this was shimming Radium in tests when we don't want to anymore.

File: apps/test/util/testUtils.js [hunk 5]
```diff
@@ -3,16 +3,10 @@ import {assert} from './configuredChai';
...
-var React = require('react');
-var ReactDOM = require('react-dom');
-var Radium = require('radium');
 
 exports.setExternalGlobals = function () {
-  window.React = React;
-  window.ReactDOM = ReactDOM;
   window.$ = $;
   window.jQuery = $;
-  window.Radium = Radium;
```
Comment (islemaster) [inline apps/test/util/testUtils.js:15]: Here's the second kludge we get to (mostly) remove by doing this - tests don't depend on having these on window anymore.

**Comment by islemaster** [comment]
One gotcha with this change is that you need to require React in any file that uses JSX, regardless of whether you actually use the `React` variable or not.  This is because JSX components get transpiled into `React.createElement` calls.

This also means you _must_ import react as "React" - `import AliasedReact from 'react';` isn't going to cut it.

File: apps/src/applab/designMode.js (hunk 3 above)
Comment (Bjvanminnen) [inline apps/src/applab/designMode.js:4]: yep

**Comment by Bjvanminnen** [comment]
This is awesome. Thanks for doing this!

**Comment by islemaster** [comment]
Note: It looks like this change grew the overall build output very slightly (from more require statements, I suppose) and moved a large amount of code (about ~1.5M, probably React+ReactDOM) from code-studio.js to code-studio-common.js.  This is expected, since the factored code-studio files now require those normally instead of expecting a global.

Here's a diff of debug build output before and after this change:

screenshot from 2016-05-24 17-41-35

**Comment by islemaster** [comment]
It's worth noting that moving content from code-studio.js to code-studio-common.js should have little impact on the end-user experience, since those two files [are _usually_ loaded together](https://github.com/code-dot-org/code-dot-org/blob/26f099b7c4797c28f317bef05fd624dcf8c31198/dashboard/app/views/layouts/application.html.haml#L13).  It looks like it might mean loading two copies of React [when embedding blocks](https://github.com/code-dot-org/code-dot-org/blob/26f099b7c4797c28f317bef05fd624dcf8c31198/dashboard/app/views/levels/embed_blocks.html.erb#L12), which is not great. I [expressed concern about needing code-studio-common here](https://github.com/code-dot-org/code-dot-org/pull/7413#issuecomment-199512861).  Nevertheless, I feel this is worth the temporary price of bloating embed_blocks on the way to unifying apps and code-studio.

**Comment by joshlory** [comment]
I don't know much about the build/test changes, but the globals removal LGTM.

**Pull request merged**